src/pretty.rs

```diff
@@ -0,0 +1,471 @@
+            LessThan() => allocator.text("<"),
+            GreaterThan() => allocator.text(">"),
+            GreaterOrEq() => allocator.text(">="),
+            LessOrEq() => allocator.text("<="),
+
+            Merge() => allocator.text("&"),
+
+            op => allocator.as_string(format!("%{:?}%", op).to_lowercase()),
+        }
+    }
+}
+
+impl<'a, D, A> Pretty<'a, D, A> for RichTerm
+where
+    D: NickelAllocatorExt<'a, A>,
+    D::Doc: Clone,
+    A: Clone + 'a,
+{
+    fn pretty(self, allocator: &'a D) -> DocBuilder<'a, D, A> {
+        use Term::*;
+
+        match self.as_ref() {
+            Null => allocator.text("null"),
+            Bool(v) => allocator.as_string(v),
+            Num(v) => allocator.as_string(v),
+            Str(v) => allocator.as_string(v),
+            StrChunks(chunks) => allocator
+                .text("m%\"")
+                .append(allocator.intersperse(
+                    chunks.iter().map(|c| {
+                        match c {
+                            crate::term::StrChunk::Literal(s) => allocator.as_string(s),
+                            crate::term::StrChunk::Expr(e, i) => allocator
+                                .text("%{")
+                                .append(e.to_owned().pretty(allocator))
+                                .append(allocator.text("}")),
+                        }
+                    }),
+                    allocator.line_(),
+                ))
+                .append(allocator.text("\"%m")),
+            Fun(id, rt) => {
+                let mut params = vec![id];
+                let mut rt = rt;
+                while let Fun(id, t) = rt.as_ref() {
+                    params.push(id);
+                    rt = t
+                }
+                allocator
+                    .text("fun ")
+                    .append(allocator.intersperse(
+                        params.iter().map(|p| allocator.as_string(p)),
+                        allocator.space(),
+                    ))
+                    .append(allocator.text(" =>"))
+                    .append(allocator.softline())
+                    .append(rt.to_owned().pretty(allocator).nest(2))
+            }
+            // TODO Pattern destructuring to implement.
+            FunPattern(..) => {
+                let mut params = vec![];
+                let mut rt = &self;
+                while let FunPattern(id, dst, t) = rt.as_ref() {
+                    params.push(
+                        if let Some(id) = id {
+                            allocator.as_string(id)
+                        } else {
+                            allocator.nil()
+                        }
+                        .append(allocator.nil()),
+                    );
+                    rt = t;
+                }
+                allocator
+                    .text("fun ")
+                    .append(allocator.intersperse(params, allocator.space()))
+                    .append(allocator.text(" =>"))
```

Comment:
Sometimes you use literal spaces as here, and sometimes you use `allocator.space()`. Is there a reason to not do:

```rust
.append(allocator.space())
.append(allocator.text("=>"))
```

Here?

Reply:
It's equivalent actually. Less to write ;). But you're right, the second version is more clear.
The `space()` is a simple space, writing it as a function make it easy to replace by another type of blank. Not absolutely sure but I think it's the only reason. But yes it would be more clear to use the `space` function everywhere.

Reply:
Do they follow the same rules about breaking, indentation and such? I would expect `space()` to have a specific semantics (like this can be replaced by a newline), but I don't know, I'm just asking.

Reply:
It's actualy `line()` and `softline()` wich can be either a line or a space. `space()` is only a space if I don't mistake.